frappe/integrations/doctype/authorizenet_settings/authorizenet_settings.py

```diff
@@ -0,0 +1,216 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2020, Frappe Technologies Pvt. Ltd. and contributors
+# For license information, please see license.txt
+
+"""
+# Integrating AuthorizeNet
+
+### 1. Validate Currency Support
+
+Example:
+
+	controller().validate_transaction_currency(currency)
+
+### 2. Redirect for payment
+
+Example:
+
+	payment_details = {
+		"createTransactionRequest": {
+			"merchantAuthentication": {
+				"name": "xxxxxxxxxx",
+				"transactionKey": "xxxxxxxxxxxx"
+			},
+			"refId": "123456",
+			"transactionRequest": {
+				"transactionType": "authCaptureTransaction",
+				"amount": "5",
+				"payment": {
+					"credit_card": {
+						"card_number": "XXXX-XXXX-XXXX-XXXX",
+						"expiration_date": "YYYY-MM",
+						"card_code": "XXX"
+					}
+				},
+				"lineItems": {
+					"lineItem": {
+						"itemId": "1",
+						"name": "vase",
+						"description": "Cannes logo",
+						"quantity": "18",
+						"unitPrice": "45.00"
+					}
+				}
+			}
+		}
+	}
+
+	# redirect the user to this url
+	url = controller().get_payment_url(**payment_details)
+
+### 3. On Completion of Payment
+
+Return payment status after processing the payment
+
+"""
+from __future__ import unicode_literals
+import frappe
+import imp
+import json
+import os
+import sys
+import re
+from six.moves.urllib.parse import urlencode
+from frappe.model.document import Document
+from frappe.integrations.utils import create_payment_gateway
+from frappe.utils import get_url, call_hook_method
+from frappe.utils.password import get_decrypted_password
+
+from authorizenet import apicontractsv1
+from authorizenet.apicontrollers import createTransactionController
+
+class AuthorizenetSettings(Document):
+	supported_currencies = ["USD","CAD"]
+
+	def validate_transaction_currency(self, currency):
+		if currency not in self.supported_currencies:
+			frappe.throw(_("Please select another payment method. AuthorizeNet does not support transactions in currency '{0}'").format(currency))
+	
+	def validate(self):
+		create_payment_gateway('Authorizenet')
+		call_hook_method('payment_gateway_enabled', gateway="Authorizenet")
+
+	def get_payment_url(self, **kwargs):
+		return get_url("./integrations/authorizenet_checkout?{0}".format(urlencode(kwargs)))
+
+
+@frappe.whitelist()
+def charge_credit_card(data, card_number, expiration_date, card_code):
```

Comment:
break this into smaller functions
https://github.com/frappe/erpnext/wiki/Coding-Standards#break-functionsmethods-more-than-10-lines-of-code